fine grained rbac (not fully approved but almost complete with peer/tech preview) (#7962)
open-cluster-management/rhacm-docs: secure_cluster/fine_grain_rbac_cli.adoc

[#fine-grain-rbac-cli]
= Implementing fine-grained role-based access control in the terminal (Technology Preview)

> *Technology Preview:* {acm} supports fine-grained role-based access control (RBAC). As a cluster administrator, you can manage and control permissions with the `ClusterPermission` resource, which controls permissions at the namespace level on managed clusters, as well as at the cluster level. Grant permissions to a virtual machine namespace within a cluster without granting permission to the entire managed cluster, or virtual machine. 

Reviewer: ```suggestion
*Technology Preview:* {acm} supports fine-grained role-based access control (RBAC). As a cluster administrator, you can manage and control permissions with the `ClusterPermission` resource, which controls permissions at the namespace level on managed clusters, as well as at the cluster level. Grant permissions to a virtual machine namespace within a cluster without granting permission to the entire managed cluster.
```

Reviewer: We can't grant permission to a single virtual machine, it is only on the namespace level. Granting access to a VM namespace will grant that permission to all VMs in that namespace.

> *Technology Preview:* {acm} supports fine-grained role-based access control (RBAC). As a cluster administrator, you can manage and control permissions with the `ClusterPermission` resource, which controls permissions at the namespace level on managed clusters, rather than at the cluster level. Grant permissions to a virtual machine namespace within a cluster without granting permission to the entire managed cluster, or virtual machine. 

Reviewer: ```suggestion
*Technology Preview:* {acm} supports fine-grained role-based access control (RBAC). As a cluster administrator, you can manage and control permissions with the `ClusterPermission` resource, which controls permissions at the namespace level on managed clusters, as well as at the cluster level. Grant permissions to a virtual machine namespace within a cluster without granting permission to the entire managed cluster, or virtual machine. 
```

Reviewer: Not sure if my proposed change is the best wording, but I don't think we should make it sound like ClusterPermission can ONLY do permissions at namespace level. It can also do so at cluster level. Admins will use a combination of both. However the feature itself is more about the fine grained part (namespace level), so it is right to focus on that part.

You can grant access to the following roles for {ocp-virt-short}, which are extensions of the default roles:

- `kubevirt.io:view`: only view resources
- `kubevirt.io:edit`: modify resources
- `kubevirt.io:admin`: view, modify, delete resources; grant permissions

. Enable `fine-grained-rbac-preview` in the `MultiClusterHub` resource. 

.. Run the following command:

+
[source,bash]
----
oc edit mch -n open-cluster-management multiclusterhub
----

.. Edit to change the `configOverrides` specification from `enabled: false` to `enabled: true`. See the following example with the feature enabled:

+
[source,yaml]
----
 - configOverrides: {}
      enabled: true
      name: fine-grained-rbac-preview
----

+
*Note:* Run `oc get mch -A` to get  the name and namespace of the `MultiClusterHub` resource if you do not use the `open-cluster-management` namespace.

. Label your `local-cluster` with `environment=virtualization`. Run the following command:
+
[source,bash]
----
oc label managedclusters local-cluster environment=virtualization
----

. Change the `policy-virt-clusterroles` to `enforce`, which adds the `kubevirt` `clusterroles` to the hub cluster. 

> .. Run the following command to get the policy:

Reviewer: ```suggestion
.. Run the following command to edit the policy:
```

+
[source,bash]
----

> oc get policy -n open-cluster-management-global-set policy-virt-clusterroles

Reviewer: ```suggestion
oc edit policy -n open-cluster-management-global-set policy-virt-clusterroles
```

+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: acm-vm-rbac-required <1>
rules:
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["kubevirtprojects"]
    verbs: ["list"]
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["list","get","watch"]
  - apiGroups: ["cluster.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["get"]
    resourceNames: ["cluster01", "cluster02", "cluster03"] <2>
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: acm-vm-rbac-required <3>
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: acm-vm-rbac-required <4> 
subjects:
  - kind: User <5>
    apiGroup: rbac.authorization.k8s.io
    name: user1 <6>
----
1. You can use a different name, but must continue to use that name.
2. Add the names of the managed clusters that you want your User or Group to access.

> 3. You can use any name, but this name must match the previous name that you used.

Reviewer: ```suggestion
3. You can use any name.
```

Reviewer: Number 3 name can be anything. The only names that need to match are numbers 1 and 4.

4. Ensure the name matches the `ClusterRole` name.

> 5. Choose a User or Group.

Reviewer: ```suggestion
5. Choose User or Group.
```

6. Specify a User or Group name.

. Apply the `ClusterRole` resource. Run the following command. Change the file name only if you changed it earlier in the process.

+
[source,bash]
----
oc apply -f acm-vm-rbac-required.yml
----

. Create a YAML file for the `ClusterPermission` resource and name the file `cluster01-prod-admin.yml`. 

. Assign fine-grain role-based access from the `ClusterPermission` resource by specifying the cluster name, managed cluster name, and Users or Group name:

+
[source,bash]
----
apiVersion: rbac.open-cluster-management.io/v1alpha1
kind: ClusterPermission
metadata:
  name: <cluster01-prod-admin> <1>
  namespace: <cluster01> <2>
spec:
  roleBindings:
    - name: cluster01-prod-admin <3>
      namespace: prod <4>
      roleRef:
        name: kubevirt.io:admin
        apiGroup: rbac.authorization.k8s.io
        kind: ClusterRole
      subjects:
        - kind: User <5>
          apiGroup: rbac.authorization.k8s.io
          name: user1 <6>

----

> 1. Specify the cluster name for the permissions.

Reviewer: Name can be anything here. This is just the name of the ClusterPermission object, not cluster name.

2. Specify the managed cluster namespace for permissions.

> 3. Use the cluster name for `RoleBindings`, which assigns roles to Users or Groups.

Reviewer: Name can be anything here as well. This is just the name of the RoleBinding that gets created on the managed cluster. It does not need to be the cluster name.

Author: What we normally do is `<cluster-name>`, `<cluster-namespace>`, where there are names that can be anything. We have a name here probably for testing. I think that can potentially be confusing bc we don't tell them to change it.

`<cluster01-prod-admin>` for instance -- can it be `<cluster-permission-name>`

4. Specify the namespace in the managed cluster to which the User or Group is granted access.
5. Choose User or choose Group.
6. Specify the User or Group name.

. Run the following command to apply the file. Change the name of the file if you changed the name previously:

+
[source,bash]
----
oc apply -f cluster01-prod-admin.yml
----

. *Optional:* If `observability` is enabled, create an additional `RoleBinding` on the hub cluster so that users can view virtual machine details in Grafana.

.. Create the `RoleBinding` resource for Grafana access. See the following sample YAML file with `name: user-observability-grafana-access`:

+
[source,yaml]
----
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
 name: user-observability-grafana-access

>  namespace: cluster01

Author: and this would be `<cluster-namespace>`

subjects:
 - kind: User
   apiGroup: rbac.authorization.k8s.io
   name: user1
roleRef:
 apiGroup: rbac.authorization.k8s.io
 kind: ClusterRole

>  name: view

Reviewer: We should consider adding number explanations the same way here as we did with the other yaml examples. If you choose to do so, here are the explanations:

"name: user-observability-grafana-access" - this can be any name
"namespace: cluster01" - this has to be the cluster name of which you are granting the user permissions for
"- kind: User" - either User or Group
"name: user1" - specified User or Group name

Author: Done!

----

.. Apply the `ClusterRoleBinding` resource with the following command:

+
[source,bash]
----
oc apply -f user-observability-grafana-access.yml
----

> 

Author: The CLI procedure is ready for tech review, see comments.^^^

Learn how to set up for fine-grained role-based access control (RBAC) and the `ClusterPermission` resource from the terminal.

*Required access:* Cluster administrator

To learn about {ocp-short} default and virtualization roles and permissions, see link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/virtualization/about#authorization_virt-security-policies[Authorization] in the {ocp-short} documentation.

See xref:../secure_cluster/rbac_implement_rhacm.adoc#rhacm-rbac-implement[Implementing role-based access control] for more details about {acm-short} role-based access.

.Prerequisites

See the following requirements to begin using fine-grained role-based access control:

. Your `MultiClusterHub` custom resource `spec.overrides.components` field for `search` must `enabled` to retrieve a list of managed clusters namespaces that can represent virtual machines that are used for access control.

> . You need virtual machines.

Reviewer 2: Were you intending for this to be an ordered list (steps) or unordered (bullet points)?

Author: oh, yes, these were originally steps, so I need to change them to bullets.

Author: changed this.

 
[#assign-fine-grain-rbac]
== Assigning fine-grained role-based access control in the terminal

You can grant access to the following roles for {ocp-virt-short}, which are extensions of the default roles:

- `kubevirt.io:view`: only view resources
- `kubevirt.io:edit`: modify resources
- `kubevirt.io:admin`: view, modify, delete resources; grant permissions

> 

Reviewer 2: I recommend that you add "Complete the following steps:"

. Enable `fine-grained-rbac-preview` in the `MultiClusterHub` resource. 

.. Run the following command:

+
[source,bash]
----
oc edit mch -n open-cluster-management multiclusterhub
----

.. Edit to change the `configOverrides` specification from `enabled: false` to `enabled: true`. See the following example with the feature enabled:

+

> [source,yaml]
> ----
>  - configOverrides: {}
>       enabled: true
>       name: fine-grained-rbac-preview
> ----

Reviewer 2: It seems like the indentation is incorrect here. Based on other examples it should be:
```suggestion
[source,yaml]
----
- configOverrides: {}
  enabled: true
  name: fine-grained-rbac-preview
----
```

of course ask dev to double check

Author: @mshort55 was this YAML off?

Reviewer: ```suggestion
[source,yaml]
----
    - configOverrides: {}
      enabled: true
      name: fine-grained-rbac-preview
----
```

Reviewer: She was right. If you want to make it exact with exact indentation, this is what it is.

Author: Done in another PR.

+

> *Note:* Run `oc get mch -A` to get  the name and namespace of the `MultiClusterHub` resource if you do not use the `open-cluster-management` namespace.

Reviewer 2: Removing extra space 
```suggestion
*Note:* Run `oc get mch -A` to get the name and namespace of the `MultiClusterHub` resource if you do not use the `open-cluster-management` namespace.
```

Reviewer 2: @swopebe

. Label your `local-cluster` with `environment=virtualization`. Run the following command:

> +
> [source,bash]
> ----
> oc label managedclusters local-cluster environment=virtualization
> ----

Reviewer 2: Command isn't greyed out, need a line break 
```suggestion

+
[source,bash]
----
oc label managedclusters local-cluster environment=virtualization
----
```

.. Run the following command to edit the policy:

+
[source,bash]
----
oc edit policy -n open-cluster-management-global-set policy-virt-clusterroles
----

+
.. Edit the `remediationAction` value from `inform` to `enforce`. *Important:* Two `remediationAction` specifications are in the policy, but you only need to change the later `remediationAction`. This does not apply to the first template in the YAML file. See the following sample:

+
[source,yaml]
----
  remediationAction: enforce
----

. Create the following YAML file and name the file `acm-vm-rbac-required.yml` for later steps in the procedure. You can use a different name. If you choose a different name, you need to use it throughout the procedure:

+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: acm-vm-rbac-required <1>
rules:
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["kubevirtprojects"]
    verbs: ["list"]
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["list","get","watch"]
  - apiGroups: ["cluster.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["get"]

>     resourceNames: ["cluster01", "cluster02", "cluster03"] <2>

Reviewer: ```suggestion
    resourceNames: ["<cluster01>", "<cluster02>", "<cluster03>"] <2>
```

Reviewer: I changed to variables, not sure if it is weird to have them inside quotes. Whatever clusters that customers put here need to be in quotes. Feel free to remove <> if you think its better.

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:

>   name: acm-vm-rbac-required <3>

Reviewer: ```suggestion
  name: <acm-vm-rbac-required> <3>
```

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: acm-vm-rbac-required <4> 
subjects:
  - kind: User <5>
    apiGroup: rbac.authorization.k8s.io

>     name: user1 <6>

Reviewer: ```suggestion
    name: <user1> <6>
```

----
<1> You can use a different name, but must continue to use that name.

> <2> Add the names of the managed clusters that you want your User or Group to access.

Reviewer 2: Can you explain why User and Group are capitalized? Seem like parameters and values. I recommend monospacing User and Group

Author: That's proper in the UI where they have to choose...but it's not a button or anything. The two choices are named this.

<3> You can use any name.
<4> Ensure the name matches the `ClusterRole` name.
<5> Choose User or Group.
<6> Specify a User or Group name.

> . Apply the `ClusterRole` resource. Run the following command. Change the file name only if you changed it earlier in the process.

Reviewer 2: ```suggestion
. Apply the `ClusterRole` resource. Change the file name only if you changed it earlier in the process. Run the following command:
```

Reviewer 2: @swopebe

Author: I am going to keep this as is as far as order, but will add the colon at the end. They hear "run the following command. Then they hear they need to change something in the command. Reversing order would be odd, however, you are correct that I need to add the colon at the end.

Author: colon added here.

+
[source,bash]
----
oc apply -f acm-vm-rbac-required.yml
----

> . Create a YAML file for the `ClusterPermission` resource and name the file `cluster01-prod-admin.yml`. 

Reviewer: This "cluster01-prod-admin.yml" file name can be anything also. Not sure if you want to comment that here since you did in the previous yaml file example.

> . Assign fine-grain role-based access from the `ClusterPermission` resource by specifying the cluster name, managed cluster name, and Users or Group name:

Reviewer: ```suggestion
. Assign fine-grain role-based access from the `ClusterPermission` resource by specifying the cluster name, managed cluster namespace, and Users or Group name:
```

+
[source,bash]
----
apiVersion: rbac.open-cluster-management.io/v1alpha1
kind: ClusterPermission
metadata:
  name: <cluster01-prod-admin> <1>
  namespace: <cluster01> <2>
spec:
  roleBindings:

>     - name: cluster01-prod-admin <3>

Reviewer: ```suggestion
    - name: <cluster01-prod-admin> <3>
```

>       namespace: prod <4>

Reviewer: ```suggestion
      namespace: <prod> <4>
```

      roleRef:
        name: kubevirt.io:admin
        apiGroup: rbac.authorization.k8s.io
        kind: ClusterRole
      subjects:
        - kind: User <5>
          apiGroup: rbac.authorization.k8s.io

>           name: user1 <6>

Reviewer: ```suggestion
          name: <user1> <6>
```

----
<1> Specify a name for the permissions.

> <2> Specify a namespace for permissions.

Reviewer: ```suggestion
<2> Specify a cluster name for permissions.
```

Reviewer: This part is confusing because it IS a namespace, but the namespace name is actually the cluster name for which permissions are being assigned for. It could go either way here, but I think cluster name is more appropriate. I think users will get confused here if we don't specify that it is actually the cluster name, which is a namespace on the hub.

<3> Specify a name for `RoleBindings`, which assigns roles to Users or Groups.
<4> Specify the namespace in the managed cluster to which the User or Group is granted access.
<5> Choose `User` or choose `Group`.
<6> Specify the User or Group `name`.

. Run the following command to apply the file. Change the name of the file if you changed the name previously:

+
[source,bash]
----
oc apply -f cluster01-prod-admin.yml
----

. *Optional:* If `observability` is enabled, create an additional `RoleBinding` on the hub cluster so that users can view virtual machine details in Grafana.

.. Create the `RoleBinding` resource for Grafana access. See the following sample YAML file with `name: user-observability-grafana-access`:

+
[source,yaml]
----
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:

>  name: user-observability-grafana-access <1>

Reviewer: ```suggestion
 name: <user-observability-grafana-access> <1>
```

Reviewer: This can be any name. However it is referenced by name on line 170. Not sure if that matters to you, it does not to me.

>  namespace: cluster01 <2>

Reviewer: ```suggestion
 namespace: <cluster01> <2>
```

Reviewer: This can be any managed cluster name.

subjects:
 - kind: User <3>
   apiGroup: rbac.authorization.k8s.io

>    name: user1 <4>

Reviewer: ```suggestion
   name: <user1> <4>
```

roleRef:
 apiGroup: rbac.authorization.k8s.io
 kind: ClusterRole
 name: view
----
<1> Specify a name for Grafana access.

> <2> Choose the namespace from the cluster name for which you are granting the user permissions.

Reviewer: ```suggestion
<2> Choose the cluster name for which you are granting the user permissions.
```

Reviewer: Same as the previous example, I think saying choose cluster name is better here. Even though it is a namespace value, in reality it is the cluster name for which they are granting grafana access for.

. You need virtual machines.
 
[#assign-fine-grain-rbac]
== Assigning fine-grained role-based access control in the terminal

You can grant access to the following roles for {ocp-virt-short}, which are extensions of the default roles:

- `kubevirt.io:view`: only view resources
- `kubevirt.io:edit`: modify resources
- `kubevirt.io:admin`: view, modify, delete resources; grant permissions

. Enable `fine-grained-rbac-preview` in the `MultiClusterHub` resource. 

.. Run the following command:

+
[source,bash]
----
oc edit mch -n open-cluster-management multiclusterhub
----

.. Edit to change the `configOverrides` specification from `enabled: false` to `enabled: true`. See the following example with the feature enabled:

+
[source,yaml]
----
 - configOverrides: {}
      enabled: true
      name: fine-grained-rbac-preview
----

+
*Note:* Run `oc get mch -A` to get  the name and namespace of the `MultiClusterHub` resource if you do not use the `open-cluster-management` namespace.

> . Label your `local-cluster` with `environment=virtualization`. Run the following command:

Author: Why does the user do this?

Reviewer: (in response to your why is this needed question)

This is needed for the next step to work: 
"Change the `policy-virt-clusterroles` to `enforce`, which adds the `kubevirt` `clusterroles` to the hub cluster."

If the local-cluster (hub) does not have this label, then the policy-virt-clusterroles will not apply to it and the needed roles will not be added to the hub cluster. If the needed roles are not on the hub cluster, then there will be no roles available on the UI drop down menu.

Author: ok, I think the steps give that info enough.

//why do this?
+
[source,bash]
----
oc label managedclusters local-cluster environment=virtualization
----

. Change the `policy-virt-clusterroles` to `enforce`, which adds the `kubevirt` `clusterroles` to the hub cluster. 

.. Run the following command to get the policy:

+
[source,bash]
----
oc get policy -n open-cluster-management-global-set policy-virt-clusterroles
----

+
.. Edit the `remediationAction` value from `inform` to `enforce`. *Important:* Two `remediationAction` specifications are in the policy, but you only need to change the later `remediationAction`. This does not apply to the first template in the YAML file. See the following sample:

+
[source,yaml]
----
  remediationAction: enforce
----

> //I think we should actually break this out a bit to show them.

Author: It may actually be better to break this out a bit to show them in the YAML--I didn't see an example of how we say this in the doc and we don't want to use directional language.

. Create the following YAML file and name the file `acm-vm-rbac-required.yml` for later steps in the procedure. You can use a different name. If you choose a different name, you need to use it throughout the procedure:

+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: acm-vm-rbac-required <1>
rules:
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["kubevirtprojects"]
    verbs: ["list"]
  - apiGroups: ["clusterview.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["list","get","watch"]
  - apiGroups: ["cluster.open-cluster-management.io"]
    resources: ["managedclusters"]
    verbs: ["get"]
    resourceNames: ["cluster01", "cluster02", "cluster03"] <2>

> ---

Author: We originally had ClusterRole and ClusterRoleBinding in two different steps, before the meetings and such. We also have this separated in the console process, but all together in this process.

Reviewer: (in response to your comment about why combine ClusterRole and ClusterRoleBinding together in the yaml for cli doc)

Putting ClusterRole and ClusterRoleBinding together in the same yaml is better/easier IMO because it simplifies the steps (reduces them by 1). Both can be applied by a single oc command. On UI steps, they have to be separate because UI itself separates them.

I don't feel super strongly about it either way, so if you would like this to be 2 steps instead of 1 then let me know and we can break it up.
